CI for GHC 8.10

Once staging is merged, will Plutarch continue to support GHC 8.10?

I think it should, because not all smart contracts will be moving over to GHC 9.2.

So, when using the `flake.nix` from [this PR](https://github.com/mlabs-haskell/plutus-scaffold/pull/17), and changing `ghcVersion` to `ghc8107` - I seem to be getting compile errors for fourmolu during `nix develop`-
```
src/Ormolu/Utils/Extensions.hs:24:1: error:
    Could not find module ‘Distribution.Utils.Path’
    Perhaps you meant
      Distribution.Utils.IOData (from Cabal-3.2.1.0)
      Distribution.Utils.MD5 (from Cabal-3.2.1.0)
      Distribution.Utils.Base62
```
is it expecting a more up to date cabal version?

--- a/flake.nix
+++ b/flake.nix
@@ -482,11 +482,14 @@
 
       packages = perSystem (system: self.flake.${system}.packages);
       checks = perSystem (system:
+        let ghc810 = ((projectFor system).flake { }).packages; # We don't run the tests, we just check that it builds.
+        in
         self.flake.${system}.checks
         // {
           formatCheck = formatCheckFor system;
           benchmark = (nixpkgsFor system).runCommand "benchmark" { } "${self.apps.${system}.benchmark.program} | tee $out";
-          ghc810 = ((projectFor system).flake { }).packages; # We don't run the tests, we just check that it builds.
+        } // {
+          "ghc810-plutarch:lib:plutarch" = ghc810."plutarch:lib:plutarch";
         }
       );
       check = perSystem (system:

--- a/flake.nix
+++ b/flake.nix
@@ -482,7 +482,7 @@
 
       packages = perSystem (system: self.flake.${system}.packages);
       checks = perSystem (system:
-        let ghc810 = ((projectFor system).flake { }).packages; # We don't run the tests, we just check that it builds.
+        let ghc810 = ((projectFor810 system).flake { }).packages; # We don't run the tests, we just check that it builds.
         in
         self.flake.${system}.checks
         // {

--- a/flake.nix
+++ b/flake.nix
@@ -423,14 +423,12 @@
           , word-array ^>= 0.1.0.0
       '';
 
-      projectFor = system:
+      projectForGhc = ghcName: system:
         let pkgs = nixpkgsFor system; in
         let pkgs' = nixpkgsFor' system; in
-        (nixpkgsFor system).haskell-nix.cabalProject' {
+        (nixpkgsFor system).haskell-nix.cabalProject' ({
           src = ./.;
-          compiler-nix-name = ghcVersion;
-          cabalProjectFileName = "cabal.project";
-          inherit cabalProjectLocal;
+          compiler-nix-name = ghcName;
           inherit extraSources;
           modules = [ (haskellModule system) ];
           shell = {
@@ -450,7 +448,12 @@
               #ps.shrinker-testing
             ];
           };
-        };
+        } // (if ghcName == ghcVersion then {
+          inherit cabalProjectLocal;
+        } else { }));
+
+      projectFor = projectForGhc ghcVersion;
+      projectFor810 = projectForGhc "ghc8107";
 
       formatCheckFor = system:
         let
@@ -483,6 +486,7 @@
         // {
           formatCheck = formatCheckFor system;
           benchmark = (nixpkgsFor system).runCommand "benchmark" { } "${self.apps.${system}.benchmark.program} | tee $out";
+          ghc810 = ((projectFor system).flake { }).packages; # We don't run the tests, we just check that it builds.
         }
       );
       check = perSystem (system:

--- a/flake.nix
+++ b/flake.nix
@@ -427,7 +427,13 @@
         let pkgs = nixpkgsFor system; in
         let pkgs' = nixpkgsFor' system; in
         (nixpkgsFor system).haskell-nix.cabalProject' ({
-          src = ./.;
+          # This is truly a horrible hack but is necessary. We can't disable tests otherwise in haskell.nix.
+          src = if ghcName == ghcVersion then ./. else
+          pkgs.runCommand "fake-src" { } ''
+            cp -rT ${./.} $out
+            chmod u+w $out $out/plutarch.cabal
+            sed -i '/-- Everything below this line is deleted for GHC 8.10/,$d' $out/plutarch.cabal
+          '';
           compiler-nix-name = ghcName;
           inherit extraSources;
           modules = [ (haskellModule system) ];
@@ -478,7 +484,9 @@
       inherit extraSources cabalProjectLocal haskellModule;
 
       project = perSystem projectFor;
+      project810 = perSystem projectFor810;
       flake = perSystem (system: (projectFor system).flake { });
+      flake810 = perSystem (system: (projectFor810 system).flake { });
 
       packages = perSystem (system: self.flake.${system}.packages);
       checks = perSystem (system:

--- a/plutarch.cabal
+++ b/plutarch.cabal
@@ -128,6 +128,8 @@ library
   if flag(development)
     cpp-options: -DDevelopment
 
+-- Everything below this line is deleted for GHC 8.10
+
 test-suite examples
   import:             c
   type:               exitcode-stdio-1.0

